src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6266,13 +6335,135 @@ private TypeSyntax ParseUnderlyingType(bool parentIsParameter, NameOptions optio
+                    finally
+                    {
+                        _pool.Free(modifiers);
+                    }
+                }
+
+                var greaterThanToken = EatToken(SyntaxKind.GreaterThanToken);
+                var funcPointer = SyntaxFactory.FunctionPointerType(@delegate, asterisk, callingConvention, lessThanToken, types, greaterThanToken);
+                funcPointer = CheckFeatureAvailability(funcPointer, MessageID.IDS_FeatureFunctionPointers);
+                return funcPointer;
+            }
+            finally
+            {
+                _termState = saveTerm;
+                _pool.Free(types);
+            }
+
+            PostSkipAction skipBadFunctionPointerParameterListTokens()
+            {
+                CSharpSyntaxNode? tmp = null;
+                Debug.Assert(types.Count > 0);
+                return SkipBadSeparatedListTokensWithExpectedKind(ref tmp,
+                    types,
+                    isNotExpectedFunction: p => p.CurrentToken.Kind != SyntaxKind.CommaToken,
+                    abortFunction: p => p.IsTerminator(),
+                    expected: SyntaxKind.CommaToken);
+            }
+        }
+
+        private bool IsFunctionPointerParameterModifier()
+            => IsParameterModifier(CurrentToken.Kind) || CurrentToken.Kind == SyntaxKind.ReadOnlyKeyword;
+
+        private bool IsFunctionPointerStart()
+            => CurrentToken.Kind == SyntaxKind.DelegateKeyword && PeekToken(1).Kind == SyntaxKind.AsteriskToken;
+
+        // We attempt to make error recovery a little nicer here by only treating a single
+        // identifier token as a calling convention if the token afterwords is actually a
+        // possible start to the function pointer parameter list. The intuition is that, for
+        // some partial trees like this:
+        //
+        //  (delegate*MyProperty.MyMethod();
+        //
+        // The user is actually in the middle of adding a cast, and if we interpreted MyProperty
+        // as a calling convention, we'd get a much worse set of errors than if we treated the
+        // function pointer as having no calling convention.
+        // Also for nicer error recovery, we only treat predefined types as if they could be the
+        // calling convention. For any other keyword, they're almost certainly part of some other
+        // construct, and would produce better errors if parsed separately. The user could have this:
+        //
+        // delegate* // Declaration in progress
+        // while (true) {}
+        //
+        // The parse tree will look much better if the while(true) is considered a separate structure,
+        // even though it does look like it could be the start of an invalid function pointer definition.
+        private bool IsPossibleCallingConvention()
+            => !IsPossibleFunctionPointerParameterListStart(CurrentToken) &&
+               IsPossibleFunctionPointerParameterListStart(PeekToken(1)) &&
```

Comment:
>IsPossibleFunctionPointerParameterListStart [](start = 15, length = 43)

Why require LessThanToken after the identifier when we are at the end of the text?

Reply:
I feel like parsing the identifier at the end of the file like this is such a corner case that, so long as we don't crash and we produce _something_ halfway reasonable, I'm don't want to do anything else special. I'll add a test.

---
In reply to: [349360246](https://github.com/dotnet/roslyn/pull/39879#discussion_r349360246) [](ancestors = 349360246)

Reply:
> I feel like parsing the identifier at the end of the file like this is such a corner case that, so long as we don't crash and we produce something halfway reasonable, I don't want to do anything else special. I'll add a test.

It looks like there are examples in this file to treats ```EndOfFileToken``` specially to keep detecting certain constructs at the end of the text.  For example in ```IsPossibleAccessorModifier```:
```
            if (token.Kind == SyntaxKind.CloseBraceToken || token.Kind == SyntaxKind.EndOfFileToken)
            {
                // If we see "{ get { } public }
                // then we will think that "public" likely starts an accessor.
                return true;
            }
```

---
In reply to: [349362938](https://github.com/dotnet/roslyn/pull/39879#discussion_r349362938) [](ancestors = 349362938,349360246)

Reply:
> It looks like there are examples in this file to treats EndOfFileToken specially to keep detecting certain constructs at the end of the text.

I special case the valid calling conventions. Beyond that, I'm fine with whatever is detected.

---
In reply to: [349380710](https://github.com/dotnet/roslyn/pull/39879#discussion_r349380710) [](ancestors = 349380710,349362938,349360246)